pkg/proxy/service.go

```diff
@@ -0,0 +1,193 @@
+	// existingPorts is going to store all the unique newly merged services' identifier.
+	existingPorts := sets.NewString()
+	for svcPortName, info := range other {
+		// Take ServicePortName.String() as the newly merged service's identifier and put it into existingPorts.
+		existingPorts.Insert(svcPortName.String())
+		_, exists := (*sm)[svcPortName]
+		if !exists {
+			glog.V(1).Infof("Adding new service port %q at %s", svcPortName, info.String())
+		} else {
+			glog.V(1).Infof("Updating existing service port %q at %s", svcPortName, info.String())
+		}
+		(*sm)[svcPortName] = info
+	}
+	return existingPorts
+}
+
+// Unmerge deletes all other ServiceMap's elements from current ServiceMap.  The returned value of Merge() is the `existingPorts` argument of Unmerge().
+// If a service's identifier exists in `existingPorts`, it means it's an `Update` event and will skip delete, otherwise it's a `Delete` event and will
+// delete the service from ServiceMap.
+// We pass in the `UDPStaleClusterIP`argument for storing the stale udp service cluster IP and clear stale udp connection later.
+func (sm *ServiceMap) Unmerge(other ServiceMap, existingPorts, UDPStaleClusterIP sets.String) {
+	for svcPortName := range other {
+		// skip the delete for `Update` event.
+		if existingPorts.Has(svcPortName.String()) {
+			continue
+		}
+		info, exists := (*sm)[svcPortName]
+		if exists {
+			glog.V(1).Infof("Removing service port %q", svcPortName)
+			if info.Protocol() == api.ProtocolUDP {
+				UDPStaleClusterIP.Insert(info.ClusterIP())
+			}
+			delete(*sm, svcPortName)
+		} else {
+			glog.Errorf("Service port %q removed, but doesn't exists", svcPortName)
+		}
+	}
+}
+
+// Update updates given service's change map based on the <previous, current> service pair.  It returns true if items changed,
+// otherwise return false.  Update can be used to add/update/delete items of ServiceChangeMap.  For example,
+// Add item
+//   - pass <nil, service> as the <previous, current> pair.
+// Update item
+//   - pass <oldService, service> as the <previous, current> pair.
+// Delete item
+//   - pass <service, nil> as the <previous, current> pair.
+func (scm *ServiceChangeMap) Update(namespacedName *types.NamespacedName, previous, current *api.Service, makeServicePort func(svcPortName ServicePortName, servicePort *api.ServicePort, service *api.Service) ServicePort) bool {
+	scm.lock.Lock()
+	defer scm.lock.Unlock()
+
+	change, exists := scm.items[*namespacedName]
+	if !exists {
+		change = &serviceChange{}
+		change.previous = serviceToServiceMap(previous, makeServicePort)
+		scm.items[*namespacedName] = change
+	}
+	change.current = serviceToServiceMap(current, makeServicePort)
+	// if change.previous equal to change.current, it means no change
+	if reflect.DeepEqual(change.previous, change.current) {
+		delete(scm.items, *namespacedName)
+	}
+	return len(scm.items) > 0
+}
+
+// serviceToServiceMap translates single Service object to proxyServiceMap.
```

Comment:
"translates single Service object to proxyServiceMap" -> "translate **a** single Service object to a ServiceMap"

Reply:
It's fixed now.